Rockchip: add support for RK3588, RK3576, and RK356X devices

chewitt opened the pull request:
This supersedes #7864 which has been waiting for numerous upstream dependencies to reach mythical "minimum viable product" state for LibreELEC use. Support is still rather experimental (far too many unmerged kernel patches are needed) but a usable working state with many fewer patches is visible. The main changes in the PR are:

* Add devices for RK3588, RK3576, and RK356X (RK3566, RK3568)
* Consolidate RK kernel defconfigs into two (arm and aarch64)
* Consolidate firmware picks into one
* Cleanup boot things (as per @Kwiboo earlier changes)
* Cleanup README files to reduce maintenance
* Use a tagged kernel patchset so we know the status/version/iteration and source of all patches
* Use a tagged u-boot patchset (ditto)
* Update to a current `rkbin` version
* Add `emmctool` to RK images to simplify writing board images to internal storage
* Add `ledfix` to RK images to disable heartbeat LEDs on boot

The Kernel and U-Boot patchsets are generated from:

* https://github.com/chewitt/linux/commits/rockchip-6.16.y
* https://github.com/chewitt/u-boot/commits/rockchip-2025.07

Issues:

~RK3399 and RK3328 devices needing `rockchip_legacy_boot` fail to compile during `mkimage` so I've temp dropped them from the `uboot_helper` file. I've no idea what the issue is; would be nice if someone who better understands the RK boot plumbing can point out the problem (and solution) so this can be dropped. I've labelled 'merge blocked' for this reason.~

~On RK3576 the `rock-4d` device is currently using the `roc-pc` u-boot defconfig as support for RK3576 still looks to be rather early stage and some fiddling was needed to create a booting 2025.07 based image. I'd expect the hackery to be droppable for the 2025.10 release.~

The FFMpeg package is currently using DetlevC upstream sources directly. Once some dependent kernel UAPI things land and his patches (or patch) firms up I'll rework things to use a normal patchset (realigned with Allwinner).

Kodi needs proper handling and selection of Video/OSD planes via DRM properties. An initial attempt was submitted and merged for Kodi 22 but caused problems with x86_64 devices on general purpose distros and was reverted. There's been no attempt to rework and resubmit changes since, so some undesirable kernel and Kodi patches are currently required.

Notes: 

Reaching "minimum viable product" means RK3588 and RK356X boards should boot to the Kodi home screen, with basic 1080p and 4K resolution support, basic HDMI (PCM) audio output, and should hardware decode non-exotic encodings of H264, HEVC ~and VP9~ media. RK3576 should be the same but this SoC is still rather new and IMHO feels like it needs a bit more time for u-boot and kernel core peripheral support to stabilise.

Support for HDMI-CEC and 10-bit colourspace (a precursor for HDR) has recently been submitted to kernel mailing lists but there are too many conflicts between current patchsets (everything is inter-dependent on the kernel DRM layer) to merge everything into a single Linux 6.16.y kernel branch. So these remain "coming soon" features and as in-flight patchsets converge over the next couple of kernel cycles, we'll be able to start picking patches to include them and experiment. The current state of pass-through audio is unknown and I haven't looked into what/where things are missing. Selection of pass-through audio remains disabled via `appliance.xml` to avoid user reports of it not working.

No functional changes for RK3399, RK3328 and RK3288 devices are intended, but these have been using a mix of patches and the source and current state of some "work in progress" items from 2018-2023 is now rather unclear. The consolidation to a properly tagged kernel patchset (so we don't end up in the same state again) may cause regressions. However I don't have much older RK hardware for testing myself (and rather limited time) so we need to crowdsource discovery of the issues and then go find the fixes.

catalinii commented:
Tested rock-5b with live tv and a quick h264 1080p + h265 4k works well with minimal cpu usage

chewitt commented:
@ShigeakiAsai Thanks for testing. I've picked that patch and will include it in the next rebase once I've finished playing 'spot the difference' with the revised patches that @Kwiboo pushed.

NB: I'm particularly keen for people to test older hardware (RK3399, RK3328, RK3288) as the few boards I have are at the bottom of a huge case that I really don't want to open (worse than pandora's box).

chewitt commented:
The `rockchip_legacy_boot` issue is resolved and RK356X workaround dropped after comparing with @Kwiboo updated patches so I've updated and pushed a rebase. I'll look into u-boot patches tomorrow.

Kwiboo commented:
@chewitt Thanks for updating!, there was a few things missing, I have pushed a rebased copy of your branch with some commit re-ordering to not break build until the final linux commit was added, please see my [chewitt-rockchip](https://github.com/Kwiboo/LibreELEC.tv/commits/chewitt-rockchip) branch.

Short summary of the diff:
- Update `rkbin` to latest `linux-6.1-stan-rkr6` SDK release (20250620)
- Drop `CONFIG_TOOLS_MKEFICAPSULE` from `bootloader/config`
- Code style cleanup of `bootloader/install`
- Remove `devices/RK3288/bootloader/config`
- Add support for `rk3566-odroid-m1s` and `rk3588s-rock-5c`
- Add `earlycon` to `EXTRA_CMDLINE`

Please also note that this PR will basically drop support for HDR/10-bit for RK3328/RK3399 (dropping of drm patches), make HW decoding more unstable on RK3328/RK3399 (soft/hard reset of rkvdec is not implemented in patches on mailing list) and will most likely break HW decoding on RK3288 for all codecs beside HEVC (dropping of iommu workaround).

knaerzche commented:
@chewitt I haven't check much, but if you're using a common kernel-config for all aarch64 SoCs
RK3328 (at least) also needs: `CONFIG_DRM_LIMA`  and `CONFIG_PHY_ROCKCHIP_INNO_HDMI` and RK3399 (at least) also needs `CONFIG_PHY_ROCKCHIP_PCIE` and `CONFIG_PHY_ROCKCHIP_EMMC`. Please re-compare the to current configs - there might be more

chewitt commented:
@Kwiboo thanks for the updates. I've reset against your branch and then reordered the Linux commit to the top as that's the one I'm most frequently revising (so can do simple `git commit --amend` with updates). I also squashed the rbin updates.

NB: For the rkbin update I previously picked the HEAD commit of the PKG_URL repo. The commit that you are using is not in that repo (but it's in your rkbin2 repo) so we're using GitHub fork magic. For package tracking to work we need to be using the right PKG_URL, so where are you sourcing from?

Thanks for the Linux patch explanations, I'll have a look at importing them as individual commits.

chewitt commented:
After quite some fiddling with `git am` and `patch -p1` this morning I've been able wrangle the `rockchip-old` patchset into a single kernel branch as individual commits: https://github.com/chewitt/linux/commits/rockchip-6.16.y-old and have tagged commits with e.g. `WIP/1001` to denote which series they were imported from. 

So to keep things moving forwards in the short-term I've reworked kernel patches into two patchsets:

* `rockchip-old` is the existing patchset for RK3288, RK3328, and RK3399 (regenerated)
* `rockchip` is the new patchset for RK356X, RK3576, and RK3588

The mid-term goal will be to review the patchset with @Kwiboo and @knaerzche to better understand status of larger bits of work, tag things better, and send any low-hanging fruit upstream. The long-term goal will be to combine the patchsets and of course get more stuff upstream. NB: The v2 HEVC patches on the mailing list have dependencies on other patch series that landed recently so I've left them as-is for now; we can switch to v2 (or later) iterations in a future rebase/update.

@Kwiboo some `1000` series patches in the `rockchip-old` patchset e.g. https://github.com/chewitt/linux/commit/738d54702f7a72ae8590f56446610c7687c737dd need to be reworked against drm/bridge changes merged for Linux 6.17.y - I was first trying to import onto a Linux 6.17-rc3 branch and ran into merge conflicts that require more than minor adjustment.

@ShigeakiAsai I've added those two RK3588 devices.

chewitt commented:
@ShigeakiAsai at this stage I'm only interested in packaging issues so we can get something merged. Feel free to point me towards fixes if you find them, but if there are none the issues are best reported to the upstream developers who wrote the driver code (not me). Given the current size of the kernel patchset (large) and state (80% unmerged patches) I expect there to be plenty of bugs.

NB: Please use the built-in `cat /path/to/log | paste` function to share log URL's as downloading/unpacking/searching zip files for the 5,000'th time is annoying. Using `paste` is less work for you too.

@swever826 image build instructions are in the wiki. If you need more direction than that ask for help in the forum.

ShigeakiAsai commented:
@chewitt 
Yes, I agree with you.
My report only means to show each device status and like current phase limitation.
And these may fixed by phase progress I think.
I think It haven't to resolve these problems imediately in this phase.

@swever826 
https://github.com/ShigeakiAsai/LibreELEC.tv/releases/tag/chewitt_rockchip_001
This time I uploaded an image that I built locally.
Feel free to try it out.

ShigeakiAsai commented:
I had also test on RK3566 devices with https://github.com/LibreELEC/LibreELEC.tv/commit/76e2c6a405693fb64292ca045002fe422417866c commit.

orangepi-3b, quartz64-a, quartz64-b, radxa-zero-3 and rock-3c
(odroid-m1s is not tested because I have no this device)

All test result was almost similar as RK3568 devices. (CPU usage, H.265 video skip, H.264 video play keyboard mouse lost)
But RK3566 H.265 video playing performance was a little worse than RK3568.
(Sometime video was skipping)

There is 1 different.
quartz64-b: Can't get ethernet DHCP IP address
This might be related, I will try it later. 
https://lore.kernel.org/all/D90C2JA20RDJ.103XJHRFDLW4U@cknow.org/

rpardini commented:
built from 12c4579bd for `nanopc-t6`, tested on T6-LTS hardware.
looking very good: h264 and h265 hw decode; CEC poweroff works. gui is snappy at 4k. type-a usb ports work (kbd, mceusb, snd_usb_audio)
kudos -- this is already enough to replace my RPi4b LE
---
I realized after testing I've Armbian's 2025.04 "edge" u-boot in SPI -- possibly with different blobs -- is it worth wiping mtd for testing LE's u-boot patches and blobs? (looks mostly rk3576-related to me)

chewitt commented:
@rpardini full wipe prob. not required for a 3588 board, it seems solid enough.
